// ext-grunt-basebuild/initConfig/compress.js
module.exports = function exports(grunt) {
    'use strict';

    var srclist = ['**/*'];
    var exclusionlist = this.learn('initConfig.buildinfo.packagefiles.exclude_from_package');
    if (exclusionlist != undefined){
        srclist = srclist.concat(exclusionlist);
    }

    var packagepostfix = "";
    if (grunt.option('withsample')) {
        packagepostfix = "sample"
    }

    return {
        package: {
            options: {
                mode: 'tgz',
                archive: '<%= buildinfo.name %>-<%= buildinfo.version%>' + packagepostfix + '-<%=buildinfo.buildNumber%>.<%= buildinfo.extension%>'
            },
            expand: true,
            cwd: 'stage',
            src: srclist,
            dest: '<%= buildinfo.name %>'
        },
        sharedpkg_ui: {
            options: {
                mode: 'tgz',
                archive: "<%= buildinfo.name %>-<%= buildinfo.version%>-<%=buildinfo.buildNumber%>.<%= buildinfo.extension%>"
            },
            files: [{
                expand: true,
                cwd: 'build',
                src: ['**']
            }]
        },
        sharedpkg_lib: {
            options: {
                mode: 'tgz',
                archive: "<%= buildinfo.name %>-<%= buildinfo.version%>-<%=buildinfo.buildNumber%>.<%= buildinfo.extension%>"
            },
            files: [{
                expand: true,
                cwd: 'build',
                src: ['**']
            }]
        },
        sharedpkg_api: {
            options: {
                mode: 'tar',
                archive: "dist/<%=buildinfo.name%>.tar"
            },
            files: [
                { expand: true, cwd: 'stage/<%=buildinfo.name%>', src: ['bin/**'], dest: '/' },
                { expand: true, cwd: 'stage/<%=buildinfo.name%>', src: ['default/**'], dest: '/' },
                { expand: true, cwd: 'stage/<%=buildinfo.name%>', src: ['tests/**'], dest: '/' }
            ]
        },
        package_app_suite: {
            options: {
                mode: 'tgz',
                archive: '<%= buildinfo.name %>-<%= buildinfo.version%>' + packagepostfix + '-<%=buildinfo.buildNumber%>.<%= buildinfo.extension%>'
            },
            expand: true,
            cwd: 'stage',
            src: srclist,
            dest: '.'
        },
        package_sample_app: {
            options: {
                mode: 'tgz',
                archive: 'app-<%= buildinfo.name %>-<%= buildinfo.version%>-<%=buildinfo.buildNumber%>.<%= buildinfo.extension%>'
            },
            files: [{
                expand: true,
                cwd: 'stage/app-<%=buildinfo.name%>',
                src: ['**'],
                dest: '.'
           }]
        }
    };
};
