// ext-grunt-basebuild/initConfig/esformatter.js
module.exports = function(grunt) {
  var options = {
    preset: 'default',
    plugins: [
      'esformatter-braces',
      'esformatter-quotes',
      'esformatter-semicolons',
      'esformatter-parseint',
      'esformatter-var-each',
      'esformatter-eol-last',
      'esformatter-remove-trailing-commas',
      'esformatter-spaced-lined-comment'
    ],
    quotes: {
      type: 'single',
      avoidEscape: true
    },
    lineBreak: {
      before: {
        AssignmentExpression: -1,
        ElseIfStatementClosingBrace: 1,
        ElseStatementClosingBrace: 1,
        FunctionExpressionClosingBrace: '1',
        FunctionDeclaration: -1,
        FunctionDeclarationClosingBrace: '1',
        IfStatementClosingBrace: 1,
        ObjectExpressionClosingBrace: -1,
        Property: -1,
        ThrowStatement: -1,
        TryClosingBrace: 1
      },
      after: {
        AssignmentExpression: -1,
        ElseIfStatementOpeningBrace: 1,
        ElseStatementOpeningBrace: 1,
        FunctionExpression: -1,
        FunctionExpressionOpeningBrace: '1',
        FunctionDeclaration: -1,
        FunctionDeclarationOpeningBrace: '1',
        FunctionDeclarationClosingBrace: -1,
        IfStatementOpeningBrace: 1,
        ObjectExpressionOpeningBrace: -1,
        Property: -1,
        TryOpeningBrace: 1
      }
    },
    whiteSpace: {
      before: {
        BlockComment: -1,
        IfStatementConditionalClosing: 0,
        ObjectExpressionOpeningBrace: 1,
        ObjectExpressionClosingBrace: 1
      },
      after: {
        BlockComment: -1,
        FunctionReservedWord: 1,
        IfStatementConditionalOpening: 0,
        ObjectExpressionOpeningBrace: 1
      }
    }
  };
  return {
    options: options,
    src: [
      '<%= buildinfo.name %>/**/*.js',
      'resources/**/*.js',
      'package/**/*.js',
      'test/**/*.js',
      'config/**/*.js',
      'Gruntfile.js',
      'build.config.js'
    ]
  };
};
