// ext-grunt-basebuild/initConfig/patternReplace.js
module.exports = function(grunt) {
    'use strict';

    var bldProperties = {};

    var version = this.learn('initConfig.buildinfo.version').split(".");

    bldProperties.version = {};
    bldProperties.version.build = this.learn('initConfig.buildinfo.buildNumber').toString();
    bldProperties.version.major = version[0];
    bldProperties.version.minor = version[1];
    bldProperties.version.revision = version[2];
    if (grunt.option('withsample')) {
        bldProperties.version.revision += 'sample';
    }
    bldProperties.copyright = "Copyright (C) 2005-2016 Splunk Inc. All Rights Reserved.";
    if (this.learn('initConfig.buildinfo.copyright')) {
        bldProperties.copyright = this.learn('initConfig.buildinfo.copyright');
    }
    bldProperties.package = {}
    bldProperties.package.name = this.learn('initConfig.buildinfo.name');
    bldProperties.ponydocs = {};
    bldProperties.ponydocs.shortname = this.learn('initConfig.buildinfo.ponydocs.shortname');
    if (this.learn('initConfig.buildinfo.packagefiles.main_app')){
        bldProperties.appname = this.learn('initConfig.buildinfo.packagefiles.main_app')
    }
    bldProperties.friendly = {};
    bldProperties.friendly.name = this.learn('initConfig.buildinfo.friendly.name');

    var dest = 'stage';

    return {
        dollar: {
            options: {
                prefix: "\\$\\{",
                suffix: "\\}",
                tokens: bldProperties,
            },
            expand: true,
            cwd: dest,
            dest: dest,
            src: ['**/README', '**/README/**', '**/README.txt', '**/default/*.conf', '**/app.manifest']
        },
        atrate: {
            options: {
                prefix: "\\@",
                suffix: "\\@",
                tokens: bldProperties,
            },
            expand: true,
            cwd: dest,
            dest: dest,
            src: ['**/README', '**/README/**', '**/README.txt', '**/default/*.conf', '**/app.manifest']
        }
    };
};
