// ext-grunt-basebuild/initConfig/rsync.js
module.exports = function(grunt) {
   'use strict';

   var splunkhomeconf;

   if (grunt.config.get('buildinfo.splunk_home') !== undefined) {
      splunkhomeconf = grunt.config.get('buildinfo.splunk_home');
   }

   var remotehostconf;

   if (grunt.config.get('buildinfo.remotehost') !== undefined) {
      remotehostconf = grunt.config.get('buildinfo.remotehost');
   }

   var remotehost = grunt.option('remotehost') || remotehostconf || '';
   var splunk_home = grunt.option('splunk_home') || splunkhomeconf || '/opt/splunk';
   splunk_home = splunk_home + "/etc/apps/";

   var rsynctarget = {
       options: {
           args: ["--verbose"],
           recursive: true
       },
       dist : {
           options: {
               args: ["-avzcph"],
               src: "stage/",
               dest: splunk_home
           }
       }
   };

   if (remotehost != ""){
       rsynctarget = {
           dist : {
               options: {
                   args: ["-avzcph"],
                   src: "stage/",
                   dest: splunk_home,
                   host: remotehost
               }
           }
       };
   }

   return rsynctarget;
};
