--- patches/jquery.iphone-style-checkboxes.js.org	2011-07-20 11:11:51.000000000 -0700
+++ jquery.iphone-style-checkboxes.js	2011-08-04 13:38:47.000000000 -0700
@@ -53,7 +53,7 @@
     if (!$.browser.msie) { return; }
 
     // Elements containing text should be unselectable
-    $.each([this.handle, this.offLabel, this.onLabel, this.container], function(el) {
+    $.each([this.handle, this.offLabel, this.onLabel, this.container], function(i, el) {
       $(el).attr("unselectable", "on");
     });
   },
