See http://jira.splunk.com:8080/browse/SPL-34437

*** jquery.ui.tablesorter.js.org	2010-10-12 16:42:54.000000000 -0700
--- jquery.ui.tablesorter.js	2010-10-12 16:42:40.000000000 -0700
***************
*** 793,799 ****
  	ts.addParser({
  		id: "shortDate",
  		is: function(s) {
! 			return /\d{1,2}[\/\-]\d{1,2}[\/\-]\d{2,4}/.test(s);
  		},
  		format: function(s,table) {
  			var c = table.config;
--- 793,799 ----
  	ts.addParser({
  		id: "shortDate",
  		is: function(s) {
! 			return (/\d{1,2}[\/\-]\d{1,2}[\/\-]\d{2,4}/).test(s);
  		},
  		format: function(s,table) {
  			var c = table.config;
