--- patches/underscore.js.org	2015-06-22 08:55:26.000000000 -0700
+++ underscore.js	2015-06-22 10:47:24.000000000 -0700
@@ -1329,15 +1329,11 @@
 
   });
 
-  // AMD registration happens at the end for compatibility with AMD loaders
-  // that may not enforce next-turn semantics on modules. Even though general
-  // practice for AMD registration is to be anonymous, underscore registers
-  // as a named module because, like jQuery, it is a base library that is
-  // popular enough to be bundled in a third party lib, but not be part of
-  // an AMD load request. Those cases could generate an error when an
-  // anonymous define() is called outside of a loader request.
+  // NOTE: this file has been patched so that Underscore defines itself as an
+  // anonymous AMD module.  This allows us to create a wrapper module called "underscore"
+  // that adds mixin functionality and enforces no-conflict mode.
   if (typeof define === 'function' && define.amd) {
-    define('underscore', [], function() {
+    define([], function() {
       return _;
     });
   }
